src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -334,41 +325,97 @@ func (c controller) unallocatedMachines(bmhList metal3v1alpha1.BareMetalHostList
 	return unallocatedList, nil
 }
 
-func (c controller) updateBMHs(bmhList metal3v1alpha1.BareMetalHostList, machineList *mapiv1beta1.MachineList) bool {
+func (c controller) copyBMHAnnotationsToStatus(bmh *metal3v1alpha1.BareMetalHost, provisioningExists bool) (bool, error) {
+	annotations := bmh.GetAnnotations()
+	content := []byte(annotations[metal3v1alpha1.StatusAnnotation])
+
+	if annotations[metal3v1alpha1.StatusAnnotation] == "" {
+		c.log.Infof("Skipping setting status of BMH host %s, status annotation not present", bmh.Name)
+		return false, nil
+	}
+	objStatus, err := c.unmarshalStatusAnnotation(content)
+	if err != nil {
+		c.log.WithError(err).Errorf("Failed to unmarshal status annotation of %s", bmh.Name)
+		return false, err
+	}
+	bmh.Status = *objStatus
+	if bmh.Status.LastUpdated.IsZero() {
+		// Ensure the LastUpdated timestamp in set to avoid
+		// infinite loops if the annotation only contained
+		// part of the status information.
+		t := metav1.Now()
+		bmh.Status.LastUpdated = &t
+	}
+	err = c.kc.UpdateBMHStatus(bmh)
+	if err != nil {
+		c.log.WithError(err).Errorf("Failed to update status of BMH %s", bmh.Name)
+		return false, err
+	}
+	return true, nil
+}
+
+func (c controller) unmarshalStatusAnnotation(content []byte) (*metal3v1alpha1.BareMetalHostStatus, error) {
+	bmhStatus := &metal3v1alpha1.BareMetalHostStatus{}
+	err := json.Unmarshal(content, bmhStatus)
+	if err != nil {
+		return nil, err
+	}
+	return bmhStatus, nil
+}
+
+func (c controller) updateBMHs(bmhList *metal3v1alpha1.BareMetalHostList, machineList *mapiv1beta1.MachineList) bool {
 	allUpdated := true
 
+	provisioningExists, err := c.kc.IsMetalProvisioningExists()
+	if err != nil {
+		return false
+	}
+
 	for i := range bmhList.Items {
 		bmh := bmhList.Items[i]
 		c.log.Infof("Checking bmh %s", bmh.Name)
-		annotations := bmh.GetAnnotations()
-		content := []byte(annotations[metal3v1alpha1.StatusAnnotation])
 
-		if annotations[metal3v1alpha1.StatusAnnotation] == "" && bmh.Spec.ConsumerRef != nil {
-			c.log.Infof("Skipping setting status of BMH host %s, status annotation not present", bmh.Name)
-			continue
-		}
-		allUpdated = false
-		if annotations[metal3v1alpha1.StatusAnnotation] != "" {
-			c.log.Infof("Updating annotations for bmh %s", bmh.Name)
-			objStatus, err := c.unmarshalStatusAnnotation(content)
-			if err != nil {
-				c.log.WithError(err).Errorf("Failed to unmarshal status annotation of %s", bmh.Name)
+		needsUpdate := false
+		if provisioningExists {
```

Comment:
it looks like we have same "if provisioningExists" can't we move this part to the one that is on line 388 ? or vice versa.